_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+A transaction should be aborted in case failure during processing, which means none of the messages from the transaction will be readable by transactional consumers.
+Instead, with the Kafka Streams API you can simply set `processing.guarantee=exactly_once_v2` to enable EOS on your existing topology.
+
+### Considerations
+
+Each producer must configure its own static and unique [`transactional.id`](https://kafka.apache.org/documentation/#producerconfigs_transactional.id) to enable the zombie fencing logic, which avoids message duplicates.
+The `transactional.id` is used to uniquely identify the same producer instance across process restarts, and it is associated with the producer session.
+When starting, the producer must call `initTransactions` one time, so the broker can tidy up the state associated with any previous incarnations of this producer, as identified by the `transactional.id`.
+Thereafter, any existing producers using the same `transactional.id` are considered zombies, and are fenced off as soon as they try to send data.
+
+> Before Kafka 2.6 the `transactional.id` had to be a static encoding of the input partitions in order to avoid ownership transfer between application instances during rebalances, that would invalidate the fencing logic.
+> This limitation was hugely inefficient, because an application instance couldn't reuse a single thread-safe producer instance, but had to create one for each input partition. 
+> It was fixed with by forcing the producer to send the consumer group metadata along with the offsets to commit ([KIP-447](https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics)).
+
+To guarantee message ordering, a given producer can have at most one open transaction (they are executed serially).
+Consumers with `isolation.level=read_committed` are not allowed to advance to offsets which include open transactions, while messages from aborted transactions are filtered out.
+
+The transaction coordinator is a module running inside every Kafka broker.
+Each coordinator owns a subset of the partitions in the internal `__transaction_state` topic (i.e. the partitions for which its broker is the leader).
+The coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.
+
+It is important to note that transactions are only supported inside a single Kafka cluster.
+If the transaction scope includes processors with externally observable side-effects, you would need to use an additional component such as `ChainedKafkaTransactionManager` from the SpringBoot project.
+This component is able to synchronize a database transaction with the Kafka transaction.
+
+#### Transaction metadata
+
+A control record is a special type of record that is used for internal operations within Kafka, rather than for normal message processing.
```

Comment:
You need a sentence here to explain that you're now talking about the structure of the log that a broker keeps on disk. 

> In order to implement transactions Kafka brokers need to include extra book-keeping information in logs. Information about producers and their transactions is stored in the `__transaction_state` log which is appended to by the transaction coordinator.
Within user logs in addition to the usual batches of user records, partition leaders append control records to track which batches have actually been committed and which rolled back.
Control records are exposed to Kafka applications by Kafka consumers, as they are an internal implementation detail of the transaction support.

Reply:
Thanks.